### ISSUE TITLE
Allow download option per user

### ISSUE THREAD

**Issue by Kudjo-Grand** [advplyr/audiobookshelf#14]
I had been thinking of something almost EXACTLY like this, so I was very happy to find it in unraid!!! Works great!

Is it possible to add allowing users to download through the browser UI? If it was on a per user basis, that would be even better. This would allow users who will be offline the chance to grab the audio ahead of time.

What do you think?

Either way, thank you for creating this!

**Comment by advplyr** [comment]
I agree completely, this will also be important for mobile. Most audiobooks have many files, so I was thinking of having different options to choose for downloading, like download a zip with all the files or download a single file with all the tracks merged.

I'm not sure what you mean by per-user basis.

**Comment by Kudjo-Grand** [comment]
Your idea there is clever. (Individual files or a "download all" option). I convert all my audiobooks down to a single file with chapters and metadata using this tool: https://github.com/yermak/AudioBookConverter

Per-user, at least the way I was thinking, something like this:
![image](https://user-images.githubusercontent.com/25695143/131976455-d53ed54e-796f-467e-85a4-2ac27cb113e8.png)

**Comment by advplyr** [comment]
I just released v1.0.0 with the experimental download tab. I explain how it functions in the tab:

![image](https://user-images.githubusercontent.com/67830747/132106125-155f1883-6144-430a-bab7-01000694fa91.png)


I haven't gotten to the user permissions yet because I want to build that to scale, so I'm going to plan out a data model.

If you have an android phone I could use your feedback on the beta I just put on the Google Play Store: https://play.google.com/store/apps/details?id=com.audiobookshelf.app

**Comment by Kudjo-Grand** [comment]
Your Android app: I just installed it and tested it. It was SUPER-FAST and easy to setup for me! And works well with what I've tested so far. (Android 11, Asus ROG Phone 3) Play, pause, resume, playing while app is in background, and playing while phone screen is off, playing while lock screen is in effect. Lock screen play/pause/resume also worked as expected. Is it possible to add skip forward/backward on the lockscreen controls also? Or is that an Android limitation?

Anyway. The Android APP working GREAT! I'll let you know if I find any strange/unexpected behavior.

**Comment by advplyr** [comment]
I think you were the first person to try it out, so thank you!

> Is it possible to add skip forward/backward on the lockscreen controls also? Or is that an Android limitation?

I just released an update to add those to the lockscreen. Opening the app should prompt you to update, hopefully that works.

**Comment by MangaValk** [comment]
> I think you were the first person to try it out, so thank you!
> 
> > Is it possible to add skip forward/backward on the lockscreen controls also? Or is that an Android limitation?
> 
> I just released an update to add those to the lockscreen. Opening the app should prompt you to update, hopefully that works.

I also tested your app already :) 
really like it, maybe the app can use a sync feature like plex has.

And maybe an idea, add a button to mark a book as already read. and filter on unread books :)

**Comment by Kudjo-Grand** [comment]
The download option is awesome! I'm testing the merge multiple files part of it now. I'll let you know what I find.

Two ideas (and you may have already thought of these, but I wanted to mention them, just in case):
1. Is it possible to give the option to download individual files instead of a full combination download only? This would help in the following way:
![image](https://user-images.githubusercontent.com/25695143/132171225-a34ff27d-464c-4968-9f0b-3eccaba4fef9.png)
Then we can pick which book we'd like instead of getting them all at once.

2. To make it easier to find and use, can we add a download button to the "main page" for each book entry?
Maybe here-ish?
![image](https://user-images.githubusercontent.com/25695143/132172054-1873dc3e-00d4-4d1e-b2e6-677db4708ec7.png)

**Comment by Merijeek** [comment]
Just gave the android app a quick test,  and so far looks really awesome.

Next suggestion would be adding in the download function to the app for
offline play.

**Comment by advplyr** [comment]
I appreciate all of your help testing and these suggestions, it is really coming along now.

I just pushed v1.0.6, which I think covers most of what was mentioned in this thread.
- Download individual tracks
- Download icon on audiobook page (just opens the download tab for now)
- Marking audiobooks as Read/Not Read
- Read audiobooks show a green progress bar

I'm still undecided on the best way to implement the read/not read/in progress filters. Syncing between the app and the desktop should be automatic, but I suspect this will cause issues in the future (ever used the Spotify app on mobile and desktop?).

User permissions is the next big thing to tackle, because right now all users can edit/download/delete/etc., which is not good for sharing. Permissions I'm thinking so far will include, updating, deleting, and downloading. Only the root user has access to the /config page.
Thanks again to you all!

**Comment by davidszp** [comment]
I have tested the Android app and the Web client. Apparently the position syncing between both platform works fine, I can start a book on the Desktop via the Webcliebt, play the book and then continue on my mobile device (as long as I am using my VPN/Home WIFI).

Download/offline listening on the mobile device would be the next logical step. In my view, this should include functionality to make files available offline on the device (download), a function to delete the downloaded/offline files to reclaim storage space on the device as well as the functionality to sync back the listening position state back to the central database once network connection is resumed.

Thank you very much for this interesting app. This looks very promosing.

**Comment by advplyr** [comment]
I just released [v0.8.0-beta to the play store](https://play.google.com/store/apps/details?id=com.audiobookshelf.app) with full offline support.

I anticipate a lot of bugs and compatibility issues, so I could use all of your help in testing it out. I had to make some small adjustments to the server for this update, so make sure your server is up to date.

The user interface is still rough around the edges, let me know if you have ideas on how to improve that. You will only see your downloaded audiobooks when not connected to your server.

**[Support]**

I focused on support for Android 10 and higher because the app is using [scoped storage](https://developer.android.com/training/data-storage/shared/media#scoped_storage_enabled), which means the app can only see files that the app created/downloaded itself, thus doesn't require permission from the user. Android 9 or lower may behave unexpectedly (if you are running an older OS, please let me know how it functions).

**[Downloading]**

When downloading audiobooks with multiple tracks the app will request a single merged audio file from the server, this is what "Preparing" means when you see that.
The audio file and cover art will both be stored in your phones /Audiobooks folder so it should be easy to find.

**[Syncing]**

When a connection with the server is established all of your local audiobook progress will be synced with the server based on the last update timestamp.


All app related issues and suggestions can go in the [audiobookshelf-app repository](https://github.com/advplyr/audiobookshelf-app)
